### PR TITLE
fix: Ignore whitespace around identifier with password strategy

### LINKED ISSUE #2158: Email address with leading/trailing whitespace treated as different identifier
### Preflight checklist

- [X] I could not find a solution in the existing issues, docs, nor discussions.
- [X] I agree to follow this project's [Code of Conduct](https://github.com/ory/kratos/blob/master/CODE_OF_CONDUCT.md).
- [X] I have read and am following this repository's [Contribution Guidelines](https://github.com/ory/kratos/blob/master/CONTRIBUTING.md).
- [ ] This issue affects my [Ory Cloud](https://www.ory.sh/) project.
- [ ] I have joined the [Ory Community Slack](https://slack.ory.sh).
- [ ] I am signed up to the [Ory Security Patch Newsletter](https://ory.us10.list-manage.com/subscribe?u=ffb1a878e4ec6c0ed312a3480&id=f605a41b53).

### Describe the bug

Suppose an account exists for `"foo@example.com"`, then it is not possible to log in using `"foo@example.com "` (same string but with trailing whitespace, which often happens on mobile phones with auto-completion) or `" foo@example.com"` (leading whitespace). In both cases, the error message reads `"The provided credentials are invalid, check for spelling mistakes in your password or username, email address, or phone number."`. I think whitespace around the identifier should be ignored.

Even worse, entirely new accounts can be registered for all of the following:

```json
" foo@example.com"
"  foo@example.com"
"   foo@example.com"
```

etc.

Interestingly, trying to register an account for `"foo@example.com "` (trailing space instead of leading) will fail, as it should. The error message there reads `"\"foo@example.com \" is not valid \"email\""`.

### Reproducing the bug

Create an account for `"foo@example.com"` and verify that logging into it works, then log out again.

At this point the following things will fail, although they should succeed:

* Logging in with `"foo@example.com "`
* Logging in with `" foo@example.com"`

The following will work, although it should fail:

* Creating an account for `" foo@example.com"`

The following will correctly fail:

* Creating an account for `"foo@example.com "`

(Note that the quotation marks are not part of user input. I had to add them to deal with Markdown formatting issues.)

### Relevant log output

_No response_

### Relevant configuration

```yml
{
  "$id": "https://schemas.ory.sh/presets/kratos/quickstart/email-password/identity.schema.json",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "Person",
  "type": "object",
  "properties": {
    "traits": {
      "type": "object",
      "properties": {
        "email": {
          "type": "string",
          "format": "email",
          "title": "E-Mail",
          "minLength": 3,
          "ory.sh/kratos": {
            "credentials": {
              "password": {
                "identifier": true
              }
            },
            "verification": {
              "via": "email"
            }
          }
        },
        "name": {
          "type": "string",
          "title": "Full name"
        }
      },
      "required": [
        "email",
        "name"
      ],
      "additionalProperties": false
    }
  }
}
```

### Version

v0.8.0-alpha.3

### On which operating system are you observing this issue?

Linux

### In which environment are you deploying?

Kubernetes with Helm

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by meyfa**
All identifiers are trimmed before they are passed to the database.

## Related issue(s)

closes #2158

## Checklist

- [X] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [X] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [X] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [X] I have read the [security policy](../security/policy).
- [X] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security. vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [X] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](docs/docs).

### CODE REVIEW

File: selfservice/strategy/password/login_test.go [hunk 1]
```diff
@@ -677,6 +677,24 @@ func TestCompleteLogin(t *testing.T) {
...
+		browserClient := testhelpers.NewClientWithCookies(t)
+		f := testhelpers.InitializeLoginFlowViaBrowser(t, browserClient, publicTS, false, false)
+
+		values := url.Values{"method": {"password"}, "password_identifier": {"  " + identifier + "  "}, "password": {pwd}, "csrf_token": {x.FakeCSRFToken}}.Encode()
+
+		_, res := testhelpers.LoginMakeRequest(t, false, false, f, browserClient, values)
+		assert.EqualValues(t, http.StatusOK, res.StatusCode)
+
+		f = testhelpers.InitializeLoginFlowViaBrowser(t, browserClient, publicTS, true, false)
+		body2, res := testhelpers.LoginMakeRequest(t, false, false, f, browserClient, values)
```
Comment (zepatrik) [inline selfservice/strategy/password/login_test.go:693]: So you are making 2 requests here, but don't really use the first one? I mean, the values are the same in both requests.

File: selfservice/strategy/password/registration_test.go [hunk 2]
```diff
@@ -587,43 +587,77 @@ func TestRegistration(t *testing.T) {
...
+			t.Run("case=leading whitespace", func(t *testing.T) {
+				testWithTransform(t, "leading", func(v url.Values) {
+					v.Set("traits.username", "  "+v.Get("traits.username"))
+				})
+			})
+
+			t.Run("case=trailing whitespace", func(t *testing.T) {
+				testWithTransform(t, "trailing", func(v url.Values) {
+					v.Set("traits.username", v.Get("traits.username")+"  ")
+				})
```
Comment (zepatrik) [inline selfservice/strategy/password/registration_test.go:660]: So you added the two transforms that add whitespace in the beginning and end, but there is no more test without adding whitespace? Can you also add a noop transform to keep the current test behavior?

**Review by zepatrik — CHANGES_REQUESTED** [review]
Thanks, this looks very good already. Just some minor test improvements and we should be ready to go :wink:

File: selfservice/strategy/password/registration_test.go (hunk 2 above)
Comment (meyfa) [reply selfservice/strategy/password/registration_test.go:660]: The no-op case still exists of course, it can be found starting at line 639. I apologize if this setup is too confusing, it was the best I could come up with to not duplicate lots and lots of code...

File: selfservice/strategy/password/login_test.go (hunk 1 above)
Comment (meyfa) [reply selfservice/strategy/password/login_test.go:693]: You are right, that is a bit weird. Honestly I just copy-pasted the capitalization test case from above and changed the identifier transformation. I think the capitalization test case might originate from `"should be a new session with refresh flag"` which would explain why there are two requests.

I can push a simplified version of both my new test case and the one above (as it suffers from the same problem), if you like?

Comment (zepatrik) [reply selfservice/strategy/password/login_test.go:693]: Yeah, makes sense.

File: selfservice/strategy/password/registration_test.go (hunk 2 above)
Comment (zepatrik) [reply selfservice/strategy/password/registration_test.go:660]: I see, sorry. It is a good approach, the diff is just hard to read.

**Review by zepatrik — APPROVED** [review]

**Review by aeneasr — APPROVED** [review]
